File: jianzhioffer/firstround/jianzhi_59_2.py
"""
方法一：遍历队列求最大值，时间复杂度O(n)，入队(list.append())时间复杂度O(1)，
出队时间复杂度O(n)(list.pop()时间复杂度为O(1)，list.pop(i)的时间复杂度为O(n))
"""
# class MaxQueue(object):
#
#     def __init__(self):
#         self.queue = []
#
#     def max_value(self):
#         """
#         :rtype: int
#         """
#         return max(self.queue) if self.queue else -1
#
#     def push_back(self, value):
#         """
#         :type value: int
#         :rtype: None
#         """
#         self.queue.append(value)
#
#     def pop_front(self):
#         """
#         :rtype: int
#         """
#         return self.queue.pop(0) if self.queue else -1


"""
方法二：维护一个单调递减的队列
"""
import queue
import collections
class MaxQueue(object):

    def __init__(self):
        self.queue = queue.Queue()  # 辅助队列
        self.deque = collections.deque()  # 单调递减队列

    def max_value(self):
        """
        :rtype: int
        """
        return self.deque[0] if self.deque else -1

    def push_back(self, value):
        """
        :type value: int
        :rtype: None
        """
        while self.deque and self.deque[-1] < value:
            self.deque.pop()  # 删除队尾元素
        self.queue.put(value)
        self.deque.append(value)

    def pop_front(self):
        """
        :rtype: int
        """
        if not self.deque:  # 如果判断self.queue是否为空，会超时
            return -1
        head = self.queue.get()
        if head == self.deque[0]:
            self.deque.popleft()
        return head


if __name__ == '__main__':
    obj = MaxQueue()
    obj.push_back(1)
    obj.push_back(2)
    param_1 = obj.max_value()
    param_2 = obj.pop_front()
    param_3 = obj.max_value()
    print("Done")

File: jianzhioffer/secondround/jianzhi_59_2.cpp
#include <deque>
#include <queue>
using namespace std;

class Solution
{
public:
	Solution():q(), aux_q() {}
	int max_value();
	void push_back(int value);
	int pop_front();

private:
	queue<int> q;
	deque<int> aux_q;
};

int Solution::max_value()
{
	if(q.empty()) return -1;
	return aux_q.front();
}

int Solution::pop_front()
{
	if(q.empty()) return -1;
	int tmp=q.front();
	if(tmp==aux_q.front()) aux_q.pop_front();
	q.pop();
	return tmp;
}

void Solution::push_back(int value)
{
	while(!aux_q.empty() && aux_q.back()<value) aux_q.pop_back();

	q.push(value);
	aux_q.push_back(value);
	return;
}

int main()
{
	Solution obj;
	obj.push_back(1);
	obj.push_back(2);
	int p1 = obj.max_value();
	int p2 = obj.pop_front();
	int p3 = obj.max_value();
	return 0;
}
